src/main/java/com/google/sps/servlets/ManageListServlet.java

```diff
@@ -0,0 +1,89 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.sps.data.User;
+import java.io.IOException;
+import java.util.List;
+import java.util.Collections;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet responsible for adding recipe keys to user's planner/cookbook lists. */
+@WebServlet("/manage-list")
+public class ManageListServlet extends HttpServlet {
+
+  // checks whether the recipe already exists in the user's list
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // gets values from request
+    long id = Long.parseLong(request.getParameter("id"));
+    String idToken = request.getParameter("idToken");
+    String type = request.getParameter("type");
+
+    boolean contains = false;
+
+    // creates a user based on the passed in user id token
+    User user = new User(idToken);
+    // gets the appropriate list based on the passed in list type
+    List<Key> keys;
+    if (type.equals("cookbook")) {
+      keys = user.getCookbookList();
+    } else if (type.equals("planner")) {
+      keys = user.getPlannerList();
+    } else {
+      keys = Collections.emptyList();
+      System.out.println("AddToListServlet: invalid type");
+    }
+
+    // creates a key based on the passed in recipe ID
+    Key key = KeyFactory.createKey("Recipe", id);
+    // checks whether the key already exists in list
+    if (keys.contains(key)) {
+      contains = true;
+    } else {
+      contains = false;
+    }
+
+    response.setContentType("text/html;");
+    response.getWriter().println(String.valueOf(contains));  
+  }
+
+  // adds a recipe key to user's list
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // gets values from request
+    long id = Long.parseLong(request.getParameter("id"));
+    String idToken = request.getParameter("idToken");
+    String type = request.getParameter("type");
+    String action = request.getParameter("action");
+
+    // creates a key based on the passed in recipe ID
+    Key key = KeyFactory.createKey("Recipe", id);
+    // creates a user based on the passed in user id token
+    User user = new User(idToken);
+
+    // adds created key to appropriate list based on passed in list type
+    if (type.equals("cookbook")) {
```

Comment:
Could you simplify this by doing something like 
`if ((type.equals("cookbook") || type.equals("planner")) && (action.equals("add") || action.equals("remove")) {do stuff} else { error message}`
This would definitely reduce the number of SOPs you have to do.
Then for the do stuff bracket you could do `if (action.equals("add")){ if (type.equals("planner")) {add planner} else {add cookbook}}`
Then for the remove you do something similar. This last suggestion is definitely out there, but I do think that doing your checks for invalid input at the beginning would help.

Reply:
I agree the if-statements are sort of a lot to look at, but I actually think there's value in having more specific error messages, if you already have to do all these checks anyway.

One idea to simplify the code would be to have a couple helper methods, something like `actionOnCookbook(key, action)` and a similar one for planner. Then this code would look like:
```
if (type.equals("planner")) {
   actionOnPlanner(key, action);
} else if (type.equals("cookbook")) {
   actionOnCookbook(key, action);
} else {
   // print "wrong type" error
}
```
and then the inner if statements would be in the helpers. You could also split it up the opposite way, and have `addToListType(key, type)` and `removeFromListType(key, type)`. These aren't huge wins but it would reduce the nesting which can help with readability.

Reply:
will do!